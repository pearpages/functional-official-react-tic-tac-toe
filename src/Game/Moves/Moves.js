import React from 'react'
import PropTypes from 'prop-types'

import { getDescription } from './moves.helpers'

import './Moves.css'

export function Moves({ history, setMove }) {
  return (<ol data-testid="Moves" className="moves">{history.map((_, index) => {
    return (
      <li key={index}>
        <button onClick={() => setMove(index)}>{getDescription(index)}</button>
      </li>
    );
  })}</ol>)
}

Moves.propTypes = {
  history: PropTypes.array.isRequired,
  setMove: PropTypes.func.isRequired
}

Moves.defaultProps = {
  history: [],
  setMove: () => undefined
}
