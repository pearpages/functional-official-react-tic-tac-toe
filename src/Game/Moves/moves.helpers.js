export function getDescription(move) {
  return (move
    ? 'Go to move #' + move
    : 'Go to game start')
}
