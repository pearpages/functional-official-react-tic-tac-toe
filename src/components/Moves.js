import React from 'react'

export function Moves({ history, setMove }) {
  return (<ol>{history.map((_, move) => {
    return (
      <li key={move}>
        <button onClick={() => setMove(move)}>{getDescription(move)}</button>
      </li>
    );
  })}</ol>)
}

function getDescription(move) {
  return (move
    ? 'Go to move #' + move
    : 'Go to game start')
}