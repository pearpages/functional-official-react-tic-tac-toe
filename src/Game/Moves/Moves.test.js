import { render, fireEvent } from '@testing-library/react'

import { Moves } from './Moves'
import { getDescription } from './moves.helpers'

describe("<Moves>", () => {
  test('When Moves does not have any props is just an empty list', () => {
    const result = render(<Moves />)
    expect(result.queryByTestId('Moves').innerHTML).toEqual('')
  })
  test('When Moves has a history prop renders the same number of buttons than elements', () => {
    const result = render(<Moves history={['e1', 'e2', 'e3', 'e4']}/>)
    expect(result.queryAllByRole('button').length).toBe(4)
  })
  test('When Moves html snapshot changes', () => {
    const result = render(<Moves history={['e1', 'e2', 'e3', 'e4']}/>)
    expect(result.getByTestId('Moves')).toMatchSnapshot()
  })
  test('When Moves has a list, should call the setMove function with the index value of that element in the history array', () => {
    const setMove =  jest.fn()
    const history = ['e1', 'e2', 'e3', 'e4']
    const result = render(<Moves history={history} setMove={setMove} />)

    const lastIndex = (history.length - 1)
    fireEvent.click(getButton(result, lastIndex))
    expect(setMove).toHaveBeenCalledWith(lastIndex)
  })
})

function getButton(result, lastIndex) {
  const textOfLastElement = getDescription(lastIndex)
  return result.getByText(textOfLastElement)
}