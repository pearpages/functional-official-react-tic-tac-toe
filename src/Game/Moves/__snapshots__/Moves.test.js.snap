// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Moves> When Moves html snapshot changes 1`] = `
<ol
  class="moves"
  data-testid="Moves"
>
  <li>
    <button>
      Go to game start
    </button>
  </li>
  <li>
    <button>
      Go to move #1
    </button>
  </li>
  <li>
    <button>
      Go to move #2
    </button>
  </li>
  <li>
    <button>
      Go to move #3
    </button>
  </li>
</ol>
`;
